File: examples/zig/benchmark.zig
//! JSON throughput through Galley's native Zig runtime: no AST, no procedures,
//! no error recovery. Parses languages/json/samples/code-01.json 50,000 times
//! on one session and reports bytes/s.

const std = @import("std");
const parser = @import("parser");

const logical_input = "languages/json/samples/code-01.json";
const default_iterations: usize = 50_000;

fn resolveInput(
    io: std.Io,
    gpa: std.mem.Allocator,
    environ_map: *std.process.Environ.Map,
    explicit: ?[]const u8,
) ![]u8 {
    if (explicit) |path| return gpa.dupe(u8, path);

    if (environ_map.get("GALLEY_CHECKOUT")) |checkout| {
        if (checkout.len > 0) {
            const candidate = try std.fs.path.join(gpa, &.{ checkout, logical_input });
            if (fileExists(io, candidate)) return candidate;
            gpa.free(candidate);
        }
    }

    const from_example = try std.fs.path.join(gpa, &.{ "..", "..", logical_input });
    if (fileExists(io, from_example)) return from_example;
    gpa.free(from_example);
    return error.MissingInput;
}

fn fileExists(io: std.Io, path: []const u8) bool {
    std.Io.Dir.cwd().access(io, path, .{}) catch return false;
    return true;
}

fn printLabeled(stdout: *std.Io.Writer, label: []const u8, value: u64) !void {
    var buffer: [32]u8 = undefined;
    try stdout.print("{s}: {s}\n", .{ label, try parser.string_utilities.formatWithThousands(value, &buffer) });
}

pub fn main(init: std.process.Init) !void {
    var stdout_buffer: [1024]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(init.io, &stdout_buffer);
    const stdout = &stdout_writer.interface;
    defer stdout.flush() catch {};

    var args = try std.process.Args.Iterator.initAllocator(init.minimal.args, init.gpa);
    defer args.deinit();
    _ = args.skip();
    const explicit = args.next();
    var iterations: usize = default_iterations;
    if (args.next()) |text| {
        iterations = std.fmt.parseInt(usize, text, 10) catch 0;
        if (iterations < 1) {
            std.debug.print("iterations must be >= 1\n", .{});
            std.process.exit(1);
        }
    }

    const path = resolveInput(init.io, init.gpa, init.environ_map, explicit) catch {
        std.debug.print("failed to read {s}\n", .{logical_input});
        std.process.exit(1);
    };
    defer init.gpa.free(path);

    const input = std.Io.Dir.cwd().readFileAlloc(init.io, path, init.gpa, .limited(std.math.maxInt(usize))) catch {
        std.debug.print("failed to read {s}\n", .{logical_input});
        std.process.exit(1);
    };
    defer init.gpa.free(input);

    const sentinel = try init.gpa.allocSentinel(u8, input.len, 0);
    defer init.gpa.free(sentinel);
    @memcpy(sentinel, input);

    var session = parser.Session.init(init.io, init.gpa, .{}) catch {
        std.debug.print("failed to create a parser session\n", .{});
        std.process.exit(1);
    };
    defer session.deinit();

    const warmup = session.parseSentinelBytes(sentinel, null) catch |err| {
        std.debug.print("warmup parse failed: {s}\n", .{@errorName(err)});
        std.process.exit(1);
    };
    if (warmup.parsed_bytes != input.len) {
        std.debug.print("warmup parse failed: parsed {d} of {d} bytes\n", .{ warmup.parsed_bytes, input.len });
        std.process.exit(1);
    }

    const start = std.Io.Clock.awake.now(init.io);
    for (0..iterations) |index| {
        const result = session.parseSentinelBytes(sentinel, null) catch |err| {
            std.debug.print("parse failed at iteration {d}: {s}\n", .{ index, @errorName(err) });
            std.process.exit(1);
        };
        if (result.parsed_bytes != input.len) {
            std.debug.print("parse failed at iteration {d}: parsed {d} of {d} bytes\n", .{
                index,
                result.parsed_bytes,
                input.len,
            });
            std.process.exit(1);
        }
    }
    const elapsed_ns: u64 = @intCast(start.durationTo(std.Io.Clock.awake.now(init.io)).toNanoseconds());
    const total: u64 = input.len * iterations;
    const bps: u64 = if (elapsed_ns == 0) 0 else total * 1_000_000_000 / elapsed_ns;

    try stdout.print("input: {s}\n", .{logical_input});
    try printLabeled(stdout, "bytes", input.len);
    try printLabeled(stdout, "iterations", iterations);
    try printLabeled(stdout, "parsed_bytes", total);
    try printLabeled(stdout, "duration_ns", elapsed_ns);
    try printLabeled(stdout, "bytes_per_second", bps);
}

File: examples/rust/src/benchmark.rs
//! JSON throughput through the Galley Rust bindings: no AST, no procedures,
//! no error recovery. Parses languages/json/samples/code-01.json 50,000 times
//! on one session and reports bytes/s.

use galley_bindings::Session;
use std::path::{Path, PathBuf};
use std::time::Instant;

const LOGICAL_INPUT: &str = "languages/json/samples/code-01.json";
const DEFAULT_ITERATIONS: usize = 50_000;

fn resolve_input(explicit: Option<&str>) -> PathBuf {
    if let Some(path) = explicit {
        return PathBuf::from(path);
    }
    if let Ok(checkout) = std::env::var("GALLEY_CHECKOUT") {
        if !checkout.is_empty() {
            let candidate = Path::new(&checkout).join(LOGICAL_INPUT);
            if candidate.is_file() {
                return candidate;
            }
        }
    }
    let from_manifest = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../../..")
        .join(LOGICAL_INPUT);
    if from_manifest.is_file() {
        return from_manifest;
    }
    Path::new("../..").join(LOGICAL_INPUT)
}

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let mut iterations = DEFAULT_ITERATIONS;
    let explicit = args.first().map(String::as_str);
    if let Some(value) = args.get(1) {
        match value.parse::<usize>() {
            Ok(count) if count >= 1 => iterations = count,
            _ => {
                eprintln!("iterations must be >= 1");
                std::process::exit(1);
            }
        }
    }

    let path = resolve_input(explicit);
    let data = match std::fs::read_to_string(&path) {
        Ok(data) => data,
        Err(_) => {
            eprintln!("failed to read {LOGICAL_INPUT}");
            std::process::exit(1);
        }
    };

    let mut session = match Session::new() {
        Ok(session) => session,
        Err(_) => {
            eprintln!("failed to create a parser session");
            std::process::exit(1);
        }
    };

    match session.parse_sentinel(&data) {
        Ok(parsed) if parsed == data.len() => {}
        Ok(parsed) => {
            eprintln!(
                "warmup parse failed: parsed {parsed} of {} bytes",
                data.len()
            );
            std::process::exit(1);
        }
        Err(error) => {
            eprintln!("warmup parse failed: {error}");
            std::process::exit(1);
        }
    }

    let start = Instant::now();
    for index in 0..iterations {
        match session.parse_sentinel(&data) {
            Ok(parsed) if parsed == data.len() => {}
            Ok(parsed) => {
                eprintln!(
                    "parse failed at iteration {index}: parsed {parsed} of {} bytes",
                    data.len()
                );
                std::process::exit(1);
            }
            Err(error) => {
                eprintln!("parse failed at iteration {index}: {error}");
                std::process::exit(1);
            }
        }
    }
    let elapsed = start.elapsed().as_nanos();
    let total = data.len() as u128 * iterations as u128;
    let bps = if elapsed == 0 {
        0
    } else {
        total * 1_000_000_000 / elapsed
    };

    println!("input: {LOGICAL_INPUT}");
    println!("bytes: {}", with_thousands(data.len() as u128));
    println!("iterations: {}", with_thousands(iterations as u128));
    println!("parsed_bytes: {}", with_thousands(total));
    println!("duration_ns: {}", with_thousands(elapsed));
    println!("bytes_per_second: {}", with_thousands(bps));
}

fn with_thousands(n: u128) -> String {
    let digits = n.to_string();
    let mut out = String::new();
    for (i, ch) in digits.chars().enumerate() {
        if i > 0 && (digits.len() - i) % 3 == 0 {
            out.push(',');
        }
        out.push(ch);
    }
    out
}
